tools/syz-declextract/syz-declextract.cpp

```diff
@@ -64,6 +106,230 @@ bool beginsWith(const std::string_view &str, const std::string_view begin) {
 
 bool contains(const std::string_view &str, const std::string_view sub) { return str.find(sub) != std::string::npos; }
 
+bool isPadding(const std::string &fieldName) {
+  return contains(fieldName, "pad") || contains(fieldName, "unused") || contains(fieldName, "_reserved");
+}
+
+const std::string makeArray(const std::string &type, const size_t len = 0) {
+  if (len == 1) {
+    return type;
+  }
+  if (len) {
+    return "array[" + type + ", " + std::to_string(len) + "]";
+  }
+  return "array[" + type + "]";
+}
+
+const std::string makeArray(const std::string &type, const size_t min, const size_t max) {
+  if (min == max) {
+    return makeArray(type, min);
+  }
+  return "array[" + type + ", " + std::to_string(min) + ":" + std::to_string(max) + "]";
+}
+
+const std::string makePtr(const std::string &dir, const std::string &type, bool isOpt = false) {
+  if (type == "void") {
+    return makePtr(dir, makeArray("int8"));
+  }
+  std::string ptr = "ptr[" + dir + ", " + type;
+  if (isOpt) {
+    return ptr + ", opt]";
+  }
+  return ptr + "]";
+}
+
+const std::string makeConst(const std::string &type, const std::string &val = "0") {
+  if (type.empty()) {
+    return "const[" + val + "]";
+  }
+  if (type == "void") {
+    return "const[" + val + ", intptr]";
+  }
+  return "const[" + val + ", " + type + "]";
+}
+
+const std::string getSyzType(const std::string &ctype, const bool isSyscallParam = false) {
+  if (!isSyscallParam && (contains(ctype, "long long") || contains(ctype, "64"))) {
```

Comment:
What happens if you remove `!isSyscallParam` here? Why is it needed?

Reply:
Some system calls are defined for 32 bit architectures, and they don't accept parameters larger than the size of an intptr.
So this ensures that in a 64-bit architecture , the size is 64 bits, and in a 32 bit architecture it's 32.

Reply:
> Some system calls are defined for 32 bit architectures, and they don't accept parameters larger than the size of an intptr.

How do we then get "long long" for them here?

Reply:
arch/arm/tools/syscall.tbl:```341  common  arm_sync_file_range     sys_sync_file_range2```
fs/sync.c:```SYSCALL_DEFINE4(sync_file_range2, int, fd, unsigned int, flags,
				 loff_t, offset, loff_t, nbytes)```
loff_t is typedef'd to long long

Reply:
OK, then we need a TODO comment here b/c this is a 32-bit syscall that accepts int64, not int32. There is probably some special arm32 calling convention for passing 64 args to functions, maybe there are passed as 2 consecutive 32-bit args or something.